source/dyaml/emitter.d

```diff
@@ -0,0 +1,1689 @@
+
+        ///Stack of states.
+        Appender!(EmitterFunction[]) states_;
+
+        ///Current state.
+        EmitterFunction state_;
+
+        ///Event queue.
+        Queue!Event events_;
+        ///Event we're currently emitting.
+        Event event_;
+
+        ///Stack of previous indentation levels.
+        Appender!(int[]) indents_;
+        ///Current indentation level.
+        int indent_ = -1;
+
+        ///Level of nesting in flow context. If 0, we're in block context.
+        uint flowLevel_ = 0;
+
+        /// Describes context (where we are in the document).
+        enum Context
+        {
+            /// Root node of a document.
+            root,
+            /// Sequence.
+            sequence,
+            /// Mapping.
+            mappingNoSimpleKey,
+            /// Mapping, in a simple key.
+            mappingSimpleKey,
+        }
+        /// Current context.
+        Context context_;
+
+        ///Characteristics of the last emitted character:
+
+        ///Line.
+        uint line_ = 0;
+        ///Column.
+        uint column_ = 0;
+        ///Whitespace character?
+        bool whitespace_ = true;
+        ///indentation space, '-', '?', or ':'?
+        bool indentation_ = true;
+
+        ///Does the document require an explicit document indicator?
+        bool openEnded_;
+
+        ///Formatting details.
+
+        ///Canonical scalar format?
+        bool canonical_;
+        ///Best indentation width.
+        uint bestIndent_ = 2;
+        ///Best text width.
+        uint bestWidth_ = 80;
+        ///Best line break character/s.
+        LineBreak bestLineBreak_;
+
+        ///Tag directive handle - prefix pairs.
+        TagDirective[] tagDirectives_;
+
+        ///Anchor/alias to process.
+        string preparedAnchor_ = null;
+        ///Tag to process.
+        string preparedTag_ = null;
+
+        ///Analysis result of the current scalar.
+        ScalarAnalysis analysis_;
+        ///Style of the current scalar.
+        ScalarStyle style_ = ScalarStyle.invalid;
+
+    public:
+        @disable int opCmp(ref Emitter);
+        @disable bool opEquals(ref Emitter);
+
+        /**
+         * Construct an emitter.
+         *
+         * Params:  stream    = Output range to write to.
+         *          canonical = Write scalars in canonical form?
+         *          indent    = Indentation width.
+         *          lineBreak = Line break character/s.
+         */
+        this(Range stream, const bool canonical, const int indent, const int width,
+             const LineBreak lineBreak) @safe
+        {
+            states_.reserve(32);
+            indents_.reserve(32);
+            stream_ = stream;
+            canonical_ = canonical;
+            nextExpected!"expectStreamStart"();
+
+            if(indent > 1 && indent < 10){bestIndent_ = indent;}
+            if(width > bestIndent_ * 2)  {bestWidth_ = width;}
```

Comment:
why are the given in values ignored in other cases? shouldn't this rather be using contracts?

Reply:
Not quite sure what is the expectation here. The parameters are not documented either :(
However this code is for the emitter, which will not be used in the foreseeable future (we would only use it if we started *emitting* YAML).
I opened an issue so this doesn't get lost: https://github.com/dlang-community/D-YAML/issues/300